lib/SILOptimizer/Differentiation/Common.cpp

```diff
@@ -452,6 +452,14 @@ findMinimalDerivativeConfiguration(AbstractFunctionDecl *original,
     auto *silParameterIndices = autodiff::getLoweredParameterIndices(
         config.parameterIndices,
         original->getInterfaceType()->castTo<AnyFunctionType>());
+
+    if (silParameterIndices->getCapacity() < parameterIndices->getCapacity()) {
+      assert(original->getCaptureInfo().hasLocalCaptures());
+      silParameterIndices =
+        silParameterIndices->extendingCapacity(original->getASTContext(),
+                                               parameterIndices->getCapacity());
+    }
+
```

Comment:
How about adding a `numCaptures` parameter to `autodiff::getLoweredParameterIndices` and make sure we allocate the right capacity in the first place?

Reply:
There are few workarounds of similar kind sprinkled over the code, so we'd need to eliminate them as well (see https://github.com/apple/swift/blob/main/lib/SILOptimizer/Mandatory/Differentiation.cpp#L518 as an example). Just setting `numCaptures" seems not work in general as we need to know how the captured arguments will be lowered.

Reply:
Yeah it's not a perfect solution.  Since captured arguments are always at the end of a SIL function, we could get the number of captures from `CaptureInfo`. Tuple captures won't be splat in SIL arguments, so we might be able to assume the number of captures in `CaptureInfo` is the number of captures in the SIL function, but I'm not sure entirely.

Reply:
Me too :) I'm planning to resolve the TODO afterwards, after checking all possibilities and fixing `getLoweredParameterIndices` properly. There is a tentative patch for this, but it needs more testing and checking.

Reply:
LGTM as is since similar workarounds already exist, but it would be really nice to fix `getLoweredParameterIndices`.  Let me know if you want to take a stab at that or rather merge this first.

Reply:
Ok sounds good.